core/modules/qos.cc

```diff
@@ -0,0 +1,441 @@
+            : AddMetadataAttr(attr, f->size, Attribute::AccessMode::kWrite);
+
+    if (f->attr_id < 0) {
+      return CommandFailure(-f->attr_id, "add_metadata_attr() failed");
+    }
+  } else {
+    return CommandFailure(EINVAL, "specify 'offset' or 'attr'");
+  }
+
+  return CommandSuccess();
+}
+
+CommandResponse Qos::Init(const bess::pb::QosArg &arg) {
+  int size_acc = 0;
+  int value_acc = 0;
+
+  for (int i = 0; i < arg.fields_size(); i++) {
+    const auto &field = arg.fields(i);
+    CommandResponse err;
+    fields_.emplace_back();
+    struct MeteringField &f = fields_.back();
+    f.pos = size_acc;
+    err = AddFieldOne(field, &f, FieldType);
+    if (err.error().code() != 0) {
+      return err;
+    }
+
+    size_acc += f.size;
+  }
+  default_gate_ = LOOKUP_FAIL_GATE;
+  total_key_size_ = align_ceil(size_acc, sizeof(uint64_t));
+  for (int i = 0; i < arg.values_size(); i++) {
+    const auto &field = arg.values(i);
+    CommandResponse err;
+    values_.emplace_back();
+    struct MeteringField &f = values_.back();
+    f.pos = value_acc;
+    err = AddFieldOne(field, &f, ValueType);
+    if (err.error().code() != 0) {
+      return err;
+    }
+
+    value_acc += f.size;
+  }
+
+  total_value_size_ = align_ceil(value_acc, sizeof(uint64_t));
+
+  uint8_t *cs = (uint8_t *)&mask;
+  for (int i = 0; i < size_acc; i++) {
+    cs[i] = 0xff;
+  }
+
+  table_.Init(total_key_size_);
+  return CommandSuccess();
+}
+
+void Qos::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
+  gate_idx_t default_gate;
+  MeteringKey keys[bess::PacketBatch::kMaxBurst] __ymm_aligned;
+  bess::Packet *pkt = nullptr;
+  default_gate = ACCESS_ONCE(default_gate_);
+
+  int cnt = batch->cnt();
+  struct value *val[cnt];
+  default_gate = ACCESS_ONCE(default_gate_);
+
+  for (const auto &field : fields_) {
+    int offset;
+    int pos = field.pos;
+    int attr_id = field.attr_id;
+
+    if (attr_id < 0) {
+      offset = field.offset;
+    } else {
+      offset = bess::Packet::mt_offset_to_databuf_offset(attr_offset(attr_id));
+    }
+
+    for (int j = 0; j < cnt; j++) {
+      char *buf_addr = batch->pkts()[j]->buffer<char *>();
+
+      /* for offset-based attrs we use relative offset */
+      if (attr_id < 0) {
+        buf_addr += batch->pkts()[j]->data_off();
+      }
+
+      char *key = reinterpret_cast<char *>(keys[j].u64_arr) + pos;
+
+      *(reinterpret_cast<uint64_t *>(key)) =
+          *(reinterpret_cast<uint64_t *>(buf_addr + offset));
+
+      size_t len = reinterpret_cast<size_t>(total_key_size_ / sizeof(uint64_t));
+
+      for (size_t i = 0; i < len; i++) {
+        keys[j].u64_arr[i] = keys[j].u64_arr[i] & mask[i];
+      }
+    }
+  }
+
+  uint64_t hit_mask = table_.Find(keys, val, cnt);
+
+  for (int init = 0; init < cnt; init++) {
+    if ((hit_mask & (1ULL << init)) == 0) {
+      EmitPacket(ctx, batch->pkts()[init], default_gate);
+      // std::cout << "qer default gate " << default_gate << std::endl;
```

Comment:
Convert to DLOG with a useful message or remove

Reply:
same for other commented out stuff

Reply:
kept logs explicitly in disabled state in Dev branch for now for better debugging logs. will be removed once we test it and ready to put in main branch